emma-language/src/main/scala/org/emmalanguage/io/csv/CSVColumn.scala

```diff
@@ -0,0 +1,132 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.emmalanguage
+package io.csv
+
+import java.{lang => jl}
+import java.{math => jm}
+import java.util.Date
+
+import scala.Function.const
+import scala.math.BigInt
+import scala.math.BigDecimal
+import scala.reflect.ClassTag
+
+/**
+ * Codec (provides encoding and decoding) type-class for CSV columns.
+ * A CSV column is an opaque value within a CSV record, encoded as a String.
+ * E.g. a CSV record of only one column would need no column delimiter.
+ *
+ * This codec is unsafe. Upon failure (mostly when decoding), an exception may be thrown.
+ * A very common exception type is [[NumberFormatException]].
+ *
+ * Usually CSV column codecs are provided for types that can be encoded as and decoded from a
+ * String, but cannot be broken down to a fixed number of fields. Otherwise [[CSVConverter]] is
+ * better suited. Instances provided by default are Scala and Java primitives, enumerations and
+ * [[Option]].
+ */
+trait CSVColumn[T] {
+
+  /**
+   * Decodes an instance of `T` from a String.
+   * Decoding is unsafe as it may throw an exception if parsing failed.
+   * A very common exception type is [[NumberFormatException]].
+   */
+  def from(column: String)(implicit csv: CSV): T
+
+  /**
+   * Encodes an instance of `T` as a String.
+   * In many cases delegating to `value.toString` is sufficient.
+   */
+  def to(value: T)(implicit csv: CSV): String
+}
+
+/** Factory methods and implicit instances of CSV column codecs. */
+object CSVColumn {
+
+  /** Summons an implicit codec for type `T` available in scope. */
+  def apply[T](implicit column: CSVColumn[T]): CSVColumn[T] = column
+
+  /** Creates a new codec for type `T` from pure conversion functions. */
+  def make[T](fromF: String => T, toF: T => String = (_: T).toString): CSVColumn[T] =
+    new CSVColumn[T] {
+      def from(column: String)(implicit csv: CSV) = fromF(column)
+      def to(value: T)(implicit csv: CSV) = toF(value)
+    }
+
+  /**
+   * Derives a new codec for type `B` from an existing codec for type `A`,
+   * given an isomorphism between `A` and `B`.
+   */
+  def iso[A, B](fromF: A => B, toF: B => A)(implicit A: CSVColumn[A]): CSVColumn[B] =
+    new CSVColumn[B] {
+      def from(column: String)(implicit csv: CSV) = fromF(A from column)
+      def to(value: B)(implicit csv: CSV) = A to toF(value)
+    }
+
+  // Scala primitives
+  implicit val stringCSVColumn:     CSVColumn[String]     = make(identity)
```

Comment:
I think this can have  its own instance which also implements quotation and escaping.

In order to amortize the cost of doing a monadic lookup on every serialization / deserialization, e.g.

```scala
val escaped = for {
  c < - csv.quote
  e <- csv.escape
} yield escape(c, e)(column)
escaped.getOrElse(column)
````

I suggest to keep a memo table of escape functions that can be looked up by a concrete `csv` instance, that way we can simplify the conversion as

```scala
escape(csv).to(toF(column))
fromF(escape(csv).from(value))
```